File: integrations/kagent/appa-kagent-adk-go/inventory.go
// The tool inventory: every name ADK can dispatch on this agent, and
// its spelling on the wire.
//
// The plugin gates a tool under a structured spelling, never under the
// bare name ADK dispatches it by: an MCP tool as mcp:<toolset>/<tool>,
// a remote agent as agent:<namespace>/<agent>, a tool the kagent
// runtime attaches itself as builtin:<name>, an out-of-band flow the
// runtime main gates as gate:<name>, and the runtime's own control
// tool as appa:execute_remedy_plan. The runtime derives the canonical
// tool and whether the call is a spawn from that spelling.
//
// The runtime main builds the inventory once at startup from the
// rendered config, so what the wire can name is fixed before the model
// runs. A call of a name outside it is refused at the gate, never
// forwarded.
//
// The inverse travels with it. The runtime names a tool back to the
// model by the spelling it received, which is not a name the model can
// call, so Despell spells it into the name ADK dispatches. The builder
// owns both directions and refuses a config whose two raw names spell
// alike, so every spelling the wire carries names one tool the model
// can call.
//
//   - An MCP entry names its tools in its tool filter, and a gated agent
//     must carry one: without it the server decides the tool list at
//     runtime, and the gate cannot name what it did not see. The toolset
//     is the first DNS label of the server host in the entry's URL, the
//     name the RemoteMCPServer resource carries in the cluster. The
//     builder refuses an endpoint outside the accepted hosts — the
//     Kubernetes service forms of that same name, and loopback — so the
//     address is a cluster service form and not an arbitrary host. It
//     establishes no more than that: the toolset is the first label
//     alone, so a service of the same name in another namespace spells
//     the same identity, and an ExternalName Service resolves an
//     accepted address to a name outside the cluster.
//   - kagent renders a remote agent's tool name as
//     <namespace>__NS__<agent> with hyphens as underscores. Both halves
//     are DNS-1123 labels, which carry no underscore, so the real names
//     come back exactly. The rendering is not injective over every name
//     a config can carry — team_a__NS__x and team-a__NS__x spell alike —
//     and the builder refuses the pair rather than lose one.
//   - The builtins come from builtins.json, the manifest pinned to the
//     kagent go module this image wraps, in groups the rendered config
//     and the runtime's environment switch on.

package appakagentadk

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

//go:embed builtins.json
var builtinsManifest []byte

// manifestLane is this image's key in the shared builtin manifest.
const manifestLane = "go"

// ControlTool is the reserved tool's spelling on the wire.
const ControlTool = "appa:execute_remedy_plan"

const namespaceMark = "__NS__"

// clusterDomain is the DNS domain a Kubernetes service name ends in.
// The toolset name is the first label of the MCP host, so a host
// outside the cluster would claim the policy identity of the
// in-cluster service of that name.
const clusterDomain = "cluster.local"

// segmentRun is one segment of a wire spelling as it stands in a
// runtime string: a run that starts and ends on a core character and
// is maximal over them.
const segmentRun = `[A-Za-z0-9_-](?:[A-Za-z0-9_.-]*[A-Za-z0-9_-])?`

// segment is one segment of a canonical tool id, as the runtime admits
// it. It is the run's own grammar anchored, so every name the
// inventory accepts is a name Despell can match back: a boundary
// period would end the run short, and the spelling could never be
// found whole.
var segment = regexp.MustCompile(`^` + segmentRun + `$`)

// spelledCandidate matches a wire spelling: two or three segments,
// class:name or class:namespace/name. The inventory alone decides
// which candidate is a spelling it gave out, and Despell replaces one
// only where the identifier continues on neither side.
var spelledCandidate = regexp.MustCompile(segmentRun + `:` + segmentRun + `(?:/` + segmentRun + `)?`)

// spellingCore holds the characters that always continue a wire
// spelling; spellingSeparator holds the ones that continue it only
// where a core character follows, so the period that ends a sentence
// closes the identifier and the period inside list.json does not.
const (
	spellingCore      = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-"
	spellingSeparator = ".:/"
)

// charAt is the byte at index, or 0 outside text.
func charAt(text string, index int) byte {
	if index < 0 || index >= len(text) {
		return 0
	}
	return text[index]
}

// continues reports whether an identifier runs on past one of its
// edges. adjacent is the byte next to the candidate and beyond the one
// past it, read away from the candidate; both are 0 at the ends of the
// text, and a byte of a multi-byte rune is neither core nor separator.
func continues(adjacent, beyond byte) bool {
	if strings.IndexByte(spellingCore, adjacent) >= 0 {
		return true
	}
	return strings.IndexByte(spellingSeparator, adjacent) >= 0 && strings.IndexByte(spellingCore, beyond) >= 0
}

func MCPSpelling(toolset, tool string) string      { return "mcp:" + toolset + "/" + tool }
func AgentSpelling(namespace, agent string) string { return "agent:" + namespace + "/" + agent }
func BuiltinSpelling(name string) string           { return "builtin:" + name }

// IsSpawn reports whether a spelled tool runs another agent: the agent: class.
func IsSpawn(spelling string) bool { return strings.HasPrefix(spelling, "agent:") }

// Inventory maps the raw ADK tool names of one agent to their wire
// spellings. The zero value spells nothing, so every call is refused.
type Inventory struct {
	spellings map[string]string
	// names is the inverse: each wire spelling, back to the name ADK
	// dispatches. BuildInventory is the only builder, and it fills both
	// directions as one: a spelling is assigned to at most one raw name,
	// so the inverse loses nothing.
	names map[string]string
}

// Spelling is the wire spelling of a raw name; false for a name outside
// the inventory.
func (i Inventory) Spelling(name string) (string, bool) {
	spelled, known := i.spellings[name]
	return spelled, known
}

func (i Inventory) Len() int { return len(i.spellings) }

// Despell is text with every wire spelling of this inventory replaced
// by the name ADK dispatches.
//
// The runtime names a tool by the spelling it was given, and the model
// dispatches another name, so runtime text that reaches the model
// passes through here first. The substitution is closed: a whole
// spelling this inventory carries is replaced, and every other byte
// stands — a spelling it never gave out included.
//
// A whole spelling is one the identifier continues on neither side, so
// mcp:demo/list inside mcp:demo/list/extra names no tool this
// inventory gave out and stands, while the period that ends the
// sentence after one is punctuation and is kept.
func (i Inventory) Despell(text string) string {
	var spelled strings.Builder
	written := 0
	for _, span := range spelledCandidate.FindAllStringIndex(text, -1) {
		start, end := span[0], span[1]
		if continues(charAt(text, start-1), charAt(text, start-2)) {
			continue
		}
		if continues(charAt(text, end), charAt(text, end+1)) {
			continue
		}
		name, gaveOut := i.names[text[start:end]]
		if !gaveOut {
			continue
		}
		spelled.WriteString(text[written:start])
		spelled.WriteString(name)
		written = end
	}
	if written == 0 {
		return text
	}
	spelled.WriteString(text[written:])
	return spelled.String()
}

// InventoryRefusalKind names why a rendered config yields no inventory.
type InventoryRefusalKind int

const (
	// UnfilteredToolset: an MCP entry declares no tool filter.
	UnfilteredToolset InventoryRefusalKind = iota
	// UnspellableName: a declared name the wire cannot spell.
	UnspellableName
	// DuplicateName: one raw name declared twice.
	DuplicateName
	// CollidingSpelling: two raw names that spell alike on the wire.
	CollidingSpelling
	// ForeignAuthority: an MCP endpoint the cluster does not serve,
	// whose host would claim the toolset name of an in-cluster service.
	ForeignAuthority
)

// InventoryRefusal is BuildInventory's error: which refusal fired, the
// config path it names, and the name at fault.
type InventoryRefusal struct {
	Kind   InventoryRefusalKind
	Path   string
	Name   string
	Detail string
}

func (r *InventoryRefusal) Error() string {
	switch r.Kind {
	case UnfilteredToolset:
		return fmt.Sprintf("%s declares no tool filter, and the gate names only what the config declares: "+
			"list under tools every tool of this server the agent may call", r.Path)
	case DuplicateName:
		return fmt.Sprintf("the config declares the tool name %q twice (%s), and the gate cannot tell the two apart: rename one of them",
			r.Name, r.Detail)
	case CollidingSpelling:
		// The detail already names both declarations and the spelling.
		return r.Detail
	default:
		return fmt.Sprintf("%s: %s", r.Path, r.Detail)
	}
}

// MCPServerSpec is one MCP entry of the rendered config.
type MCPServerSpec struct {
	Path  string
	URL   string
	Tools []string
}

// RemoteAgentSpec is one remote agent of the rendered config, under the
// name the stock builder dispatches it by.
type RemoteAgentSpec struct {
	Path string
	Name string
}

// BuiltinGroups switches the manifest groups the rendered config and
// the environment turn on. The always group needs no switch.
type BuiltinGroups struct {
	Memory     bool
	Skills     bool
	ShareTools bool
}

// InventorySpec is what the runtime main reads off the rendered config.
type InventorySpec struct {
	MCPServers   []MCPServerSpec
	RemoteAgents []RemoteAgentSpec
	Builtins     BuiltinGroups
}

// BuildInventory spells every tool the spec declares. A refusal is an
// *InventoryRefusal; a manifest this image cannot read is a plain error.
func BuildInventory(spec InventorySpec) (Inventory, error) {
	b := builder{spellings: map[string]string{}, names: map[string]string{}, sources: map[string]string{}}
	if err := b.add(ReservedTool, ControlTool, "the reserved tool"); err != nil {
		return Inventory{}, err
	}
	for _, server := range spec.MCPServers {
		if err := b.mcpServer(server); err != nil {
			return Inventory{}, err
		}
	}
	for _, remote := range spec.RemoteAgents {
		if err := b.remoteAgent(remote); err != nil {
			return Inventory{}, err
		}
	}
	groups, err := laneGroups()
	if err != nil {
		return Inventory{}, err
	}
	enabled := map[string]bool{
		"always":      true,
		"memory":      spec.Builtins.Memory,
		"skills":      spec.Builtins.Skills,
		"share_tools": spec.Builtins.ShareTools,
	}
	for group, names := range groups {
		on, known := enabled[group]
		if !known {
			return Inventory{}, fmt.Errorf("the builtin manifest names a group this image does not switch on: %s", group)
		}
		if !on {
			continue
		}
		for _, name := range names {
			if err := b.add(name, BuiltinSpelling(name), "the builtin group "+group); err != nil {
				return Inventory{}, err
			}
		}
	}
	return Inventory{spellings: b.spellings, names: b.names}, nil
}

func laneGroups() (map[string][]string, error) {
	var manifest map[string]struct {
		Groups map[string][]string `json:"groups"`
	}
	if err := json.Unmarshal(builtinsManifest, &manifest); err != nil {
		return nil, fmt.Errorf("the builtin manifest does not parse: %w", err)
	}
	lane, present := manifest[manifestLane]
	if !present {
		return nil, fmt.Errorf("the builtin manifest carries no %s lane", manifestLane)
	}
	return lane.Groups, nil
}

// builder holds both directions of one inventory, each raw name and
// each spelling taken once.
type builder struct {
	spellings map[string]string
	names     map[string]string
	sources   map[string]string
}

func (b *builder) add(name, spelling, source string) error {
	if declared, twice := b.sources[name]; twice {
		return &InventoryRefusal{Kind: DuplicateName, Path: source, Name: name, Detail: declared + " and " + source}
	}
	if spelled, taken := b.names[spelling]; taken {
		return &InventoryRefusal{Kind: CollidingSpelling, Path: source, Name: name,
			Detail: fmt.Sprintf("the config declares %q (%s) and %q (%s), and both spell as %q on the wire: "+
				"the runtime could name only one of them back to the model, so rename one of them",
				spelled, b.sources[spelled], name, source, spelling)}
	}
	b.spellings[name] = spelling
	b.names[spelling] = name
	b.sources[name] = source
	return nil
}

func (b *builder) mcpServer(server MCPServerSpec) error {
	host, hosted := hostOf(server.URL)
	toolset, spellable := toolsetOf(host)
	// A doubled underscore is the mark kagent reserves, so the runtime
	// admits no canonical id whose namespace carries one.
	if !hosted || !spellable || strings.Contains(toolset, "__") {
		return &InventoryRefusal{Kind: UnspellableName, Path: server.Path, Name: server.URL,
			Detail: fmt.Sprintf("the toolset name is the first label of the server host in the URL, and %q carries none the wire can spell", server.URL)}
	}
	if !inCluster(host) {
		return &InventoryRefusal{Kind: ForeignAuthority, Path: server.Path, Name: server.URL,
			Detail: fmt.Sprintf("%q is served outside the cluster, and its tools would claim the policy identity "+
				"mcp/%s/<tool> of the in-cluster %q: an MCP endpoint is named <service>, "+
				"<service>.<namespace>, <service>.<namespace>.svc, <service>.<namespace>.svc.cluster.local, "+
				"localhost, or 127.0.0.1",
				server.URL, toolset, toolset)}
	}
	if len(server.Tools) == 0 {
		return &InventoryRefusal{Kind: UnfilteredToolset, Path: server.Path}
	}
	for position, name := range server.Tools {
		if !segment.MatchString(name) {
			return &InventoryRefusal{Kind: UnspellableName, Path: fmt.Sprintf("%s.tools[%d]", server.Path, position), Name: name,
				Detail: fmt.Sprintf("the tool name %q is outside what the wire can spell", name)}
		}
		if err := b.add(name, MCPSpelling(toolset, name), server.Path); err != nil {
			return err
		}
	}
	return nil
}

// remoteAgent spells one remote agent by the name the entry carries.
//
// The identity is the name alone, and the entry's URL binds nothing —
// unlike an MCP entry, whose toolset is read off its own endpoint. The
// kagent controller renders both fields from the one Agent object a
// tool reference resolves to: the name from its object reference and
// the URL from toolAgentURL of that same object. The reference is a
// TypedReference (kind, name, namespace) and carries no URL, so no CRD
// can point a declared agent identity at another endpoint.
//
// Reading the identity off the URL instead would refuse two renderings
// the controller emits: a global proxy rewrites every URL to the proxy
// host and moves the real one into the x-kagent-host header, and a
// sandbox agent is reached at the controller's own address under
// /api/a2a-sandboxes/<ns>/<name>. A hand-written config.json mounted
// past the controller can still name one agent and reach another (Known
// gaps in ../IMPLEMENTATION.md).
func (b *builder) remoteAgent(remote RemoteAgentSpec) error {
	namespace, agent, marked := strings.Cut(remote.Name, namespaceMark)
	if !marked || namespace == "" || agent == "" || strings.Contains(agent, namespaceMark) {
		return &InventoryRefusal{Kind: UnspellableName, Path: remote.Path, Name: remote.Name,
			Detail: fmt.Sprintf("kagent renders a remote agent as <namespace>__NS__<agent>, and %q is not that shape", remote.Name)}
	}
	namespace, agent = strings.ReplaceAll(namespace, "_", "-"), strings.ReplaceAll(agent, "_", "-")
	if !segment.MatchString(namespace) || !segment.MatchString(agent) {
		return &InventoryRefusal{Kind: UnspellableName, Path: remote.Path, Name: remote.Name,
			Detail: fmt.Sprintf("the remote agent name %q is outside what the wire can spell", remote.Name)}
	}
	return b.add(remote.Name, AgentSpelling(namespace, agent), remote.Path)
}

// hostOf is the host of a server URL; false where it carries none.
func hostOf(raw string) (string, bool) {
	parsed, err := url.Parse(raw)
	if err != nil {
		return "", false
	}
	host := parsed.Hostname()
	if host == "" {
		return "", false
	}
	return host, true
}

// toolsetOf is the toolset name a host claims: its first label, where
// the wire can spell it.
func toolsetOf(host string) (string, bool) {
	label, _, _ := strings.Cut(host, ".")
	if !segment.MatchString(label) {
		return "", false
	}
	return label, true
}

// inCluster reports whether host is a Kubernetes service form of the
// service its first label names.
//
// The accepted forms are cluster service addresses that resolve through
// cluster DNS, and every other host is refused, so the endpoint an MCP
// entry names is a service of the cluster rather than an arbitrary
// host. It pins no single Service: the toolset is the first label
// alone, so the same name in another namespace passes, and an
// ExternalName Service resolves an accepted address to a name outside
// the cluster. <service>.<namespace> is one form short of a public
// domain name and nothing here tells the two apart, so it is accepted
// as the cluster-internal form kagent's own controller reads it as:
// isInternalK8sURL asks the API server whether that second label is a
// namespace, which this plugin cannot do.
func inCluster(host string) bool {
	host = strings.ToLower(host)
	if host == "localhost" || host == "127.0.0.1" {
		return true
	}
	labels := strings.Split(host, ".")
	switch {
	case len(labels) <= 2:
		return true
	case labels[2] != "svc":
		return false
	default:
		return len(labels) == 3 || strings.Join(labels[3:], ".") == clusterDomain
	}
}

File: integrations/kagent/appa-kagent-adk/src/appa_kagent_adk/inventory.py
"""Configured names and wire identities; MCP discovery supplies observed tools.

Optional MCP filters are not permission. The host validates authenticated metadata
before model exposure and the runtime checks every call. Full endpoint identities
never infer a provider from a hostname. Remote-agent and builtin mappings retain
kagent\'s native conventions. The inverse renders guidance in host-native names.
"""

from __future__ import annotations

import hashlib
import json
import os
import re
import string
from collections.abc import Mapping
from dataclasses import dataclass, field
from importlib import resources
from types import MappingProxyType
from typing import Any
from urllib.parse import urlsplit

from . import wire
from .config_guard import ConfigRefused

LANE = "python"
"""This image's key in the shared builtin manifest."""

SKILLS_FOLDER_ENV = "KAGENT_SKILLS_FOLDER"
"""The kagent runtime attaches its skills tools while this names a directory."""

GUIDE_ENV = "APPA_GUIDE"
GUIDE_TOOLSET = "appa-guide"
"""appa-guide's management set is served over its own endpoint and named
under one fixed toolset, so a policy names it the way it names any other
MCP tool."""

_MCP_KEYS = ("http_tools", "sse_tools")
_NAMESPACE_MARK = "__NS__"
# The characters an identifier is built of. A core character always
# continues one; a separator continues one only where a core character
# follows it, so the period that ends a sentence closes the identifier
# and the period inside ``list.json`` does not.
_CORE = frozenset(string.ascii_letters + string.digits + "_-")
_SEPARATOR = frozenset(".:/")
# One segment of a wire spelling: a run that starts and ends on a core
# character and is maximal over them.
_SEGMENT_RUN = r"[A-Za-z0-9_-](?:[A-Za-z0-9_.-]*[A-Za-z0-9_-])?"
# One segment of a canonical tool id, as the runtime admits it. It is
# the run's own grammar anchored, so every name the inventory accepts
# is a name ``despell`` can match back: a boundary period would end the
# run short, and the spelling could never be found whole.
_SEGMENT = re.compile(rf"^{_SEGMENT_RUN}$")
# The classes a wire spelling begins with. Every spelling this module
# issues carries one, so the scan below anchors on the literal class
# instead of trying a segment at every character: an unanchored scan
# backtracks over each identifier it fails to close, which is quadratic
# in one long colon-free run — and a tool result is enough to carry one.
_CLASSES = ("mcp", "agent", "builtin", "gate", wire.CONTROL_TOOL.split(":", 1)[0])
# A wire spelling as it stands in a runtime string: two or three
# segments, ``class:name`` or ``class:namespace/name``. The inventory
# alone decides which candidate is a spelling it gave out, and
# ``despell`` replaces one only where the identifier continues on
# neither side.
_SPELLED = re.compile(rf"(?:{'|'.join(_CLASSES)}):{_SEGMENT_RUN}(?:/{_SEGMENT_RUN})?")


def _char(text: str, index: int) -> str:
    """The character at ``index``, or the empty string outside ``text``."""
    return text[index] if 0 <= index < len(text) else ""


def _continues(adjacent: str, beyond: str) -> bool:
    """Whether an identifier runs on past one of its edges.

    ``adjacent`` is the character next to the candidate and ``beyond``
    the one past it, read away from the candidate; both are empty at
    the ends of the text.
    """
    return adjacent in _CORE or (adjacent in _SEPARATOR and beyond in _CORE)


def mcp_source_id(endpoint: str) -> str:
    try:
        parsed = urlsplit(endpoint)
        valid = (
            isinstance(endpoint, str)
            and parsed.scheme in ("http", "https")
            and parsed.hostname
            and not parsed.username
            and not parsed.password
            and not parsed.fragment
        )
        _ = parsed.port
    except (TypeError, ValueError):
        valid = False
    if not valid:
        raise ConfigRefused("MCP endpoint must be an HTTP(S) URL without userinfo or a fragment")
    return "server-" + hashlib.sha256(endpoint.encode("utf-8")).hexdigest()


def mcp_spelling(toolset: str, tool: str) -> str:
    return f"mcp:{toolset}/{tool}"


def agent_spelling(namespace: str, agent: str) -> str:
    return f"agent:{namespace}/{agent}"


def builtin_spelling(name: str) -> str:
    return f"builtin:{name}"


def gate_spelling(name: str) -> str:
    return f"gate:{name}"


def guide_spelling(name: str) -> str:
    return mcp_spelling(GUIDE_TOOLSET, name)


def guide_enabled(environ: Mapping[str, str] = os.environ) -> bool:
    """Whether this deployment is appa-guide's own agent.

    The inventory, which spells the management set only for the guide,
    and the toolset the guide reaches it through read the same answer,
    and a value that is neither refuses the config rather than start an
    agent that silently has no guide.
    """
    match environ.get(GUIDE_ENV, "").strip().lower():
        case "" | "false":
            return False
        case "true":
            return True
        case other:
            raise ConfigRefused(f"{GUIDE_ENV} must be true or false, not {other!r}")


def is_spawn(spelling: str) -> bool:
    """Whether a spelled tool runs another agent: the ``agent:`` class."""
    return spelling.startswith("agent:")


def builtin_manifest() -> dict[str, Any]:
    """The packaged builtin manifest, every lane included."""
    return json.loads(resources.files(__package__).joinpath("builtins.json").read_text())


@dataclass(frozen=True)
class ToolInventory:
    """The raw ADK tool names of one agent, each with its wire spelling.

    The forward direction is the whole state. The inverse is derived,
    never given, so no caller can hand in two mappings that disagree,
    and both are handed out as read-only views. A forward mapping that
    is not injective has no inverse — the runtime could name only one
    of the two raw names back to the model — and it is refused here as
    ``from_config`` refuses it at the config.
    """

    spellings: Mapping[str, str]
    names: Mapping[str, str] = field(init=False)
    """The inverse: each wire spelling, back to the name ADK dispatches."""

    def __post_init__(self) -> None:
        spellings = dict(self.spellings)
        names = {spelling: name for name, spelling in spellings.items()}
        if len(names) != len(spellings):
            raise ValueError("two raw names of the inventory spell alike, and the inverse would lose one")
        object.__setattr__(self, "spellings", MappingProxyType(spellings))
        object.__setattr__(self, "names", MappingProxyType(names))

    def spelling(self, name: str) -> str | None:
        """The wire spelling of a raw name, or None for a name outside the inventory."""
        return self.spellings.get(name)

    def despell(self, text: str) -> str:
        """``text`` with every wire spelling of this inventory replaced by the name ADK dispatches.

        The runtime names a tool by the spelling it was given, and the
        model dispatches another name, so runtime text that reaches the
        model passes through here first. The substitution is closed: a
        whole spelling this inventory carries is replaced, and every
        other byte stands — a spelling it never gave out included.

        A whole spelling is one the identifier continues on neither
        side, so ``mcp:demo/list`` inside ``mcp:demo/list/extra`` names
        no tool this inventory gave out and stands, while the period
        that ends the sentence after one is punctuation and is kept.
        """

        def replace(match: re.Match[str]) -> str:
            start, end = match.start(), match.end()
            if _continues(_char(text, start - 1), _char(text, start - 2)):
                return match.group()
            if _continues(_char(text, end), _char(text, end + 1)):
                return match.group()
            return self.names.get(match.group(), match.group())

        return _SPELLED.sub(replace, text)

    @classmethod
    def from_config(cls, config: Mapping[str, Any], environ: Mapping[str, str] = os.environ) -> ToolInventory:
        """Build the inventory of a rendered kagent config.

        Raises ``ConfigRefused`` for an invalid configured endpoint,
        a name the wire cannot spell, a raw name declared twice, and two
        raw names that spell alike.
        """
        builder = _Builder()
        builder.add(wire.RESERVED_TOOL, wire.CONTROL_TOOL, "the reserved tool")
        if guide_enabled(environ):
            for name in sorted(wire.MANAGEMENT_TOOLS):
                builder.add(name, guide_spelling(name), "the appa-guide management set")
        for key in _MCP_KEYS:
            for index, server in enumerate(_entries(config.get(key))):
                builder.mcp_server(f"{key}.{index}", server)
        for index, remote in enumerate(_entries(config.get("remote_agents"))):
            builder.remote_agent(f"remote_agents.{index}", remote)
        enabled = {
            "always": True,
            "memory": config.get("memory") is not None,
            "skills": bool(environ.get(SKILLS_FOLDER_ENV, "").strip()),
        }
        for group, names in builtin_manifest()[LANE]["groups"].items():
            if group not in enabled:
                raise ValueError(f"the builtin manifest names a group this image does not switch on: {group}")
            if enabled[group]:
                for name in names:
                    builder.add(name, builtin_spelling(name), f"the builtin group {group}")
        return cls(builder.spellings)


def _entries(raw: Any) -> list[dict[str, Any]]:
    """The dict entries of a raw config list.

    The schema validation the entrypoint ran already refused any other shape.
    """
    if not isinstance(raw, list):
        return []
    return [entry for entry in raw if isinstance(entry, dict)]


class _Builder:
    """Both directions of one inventory, each name and each spelling taken once."""

    def __init__(self) -> None:
        self.spellings: dict[str, str] = {}
        self.names: dict[str, str] = {}
        self._sources: dict[str, str] = {}

    def add(self, name: str, spelling: str, source: str) -> None:
        declared = self._sources.get(name)
        if declared is not None:
            raise ConfigRefused(
                f"the config declares the tool name {name!r} twice ({declared} and {source}), and the "
                "gate cannot tell the two apart — rename one of them"
            )
        spelled = self.names.get(spelling)
        if spelled is not None:
            raise ConfigRefused(
                f"the config declares {spelled!r} ({self._sources[spelled]}) and {name!r} ({source}), and "
                f"both spell as {spelling!r} on the wire — the runtime could name only one of them back "
                "to the model, so rename one of them"
            )
        self.spellings[name] = spelling
        self.names[spelling] = name
        self._sources[name] = source

    def mcp_server(self, path: str, server: dict[str, Any]) -> None:
        params = server.get("params")
        url = params.get("url") if isinstance(params, dict) else None
        try:
            toolset = mcp_source_id(url)
        except ConfigRefused as error:
            raise ConfigRefused(f"{path}: {error}") from None
        names = server.get("tools") or []
        if not isinstance(names, list) or not all(isinstance(name, str) for name in names):
            raise ConfigRefused(f"{path}.tools must be a list of tool names")
        for position, name in enumerate(names):
            if not _SEGMENT.match(name):
                raise ConfigRefused(
                    f"{path}.tools.{position}: the tool name {name!r} is outside what the wire can spell"
                )
            self.add(name, mcp_spelling(toolset, name), path)

    def remote_agent(self, path: str, remote: dict[str, Any]) -> None:
        """Spell one remote agent by the name the entry carries.

        The identity is the name alone, and the entry's ``url`` binds
        nothing — unlike an MCP entry, whose toolset is read off its own
        endpoint. The kagent controller renders both fields from the one
        Agent object a tool reference resolves to: the name from its
        object reference and the URL from `toolAgentURL` of that same
        object. The reference is a `TypedReference` (kind, name,
        namespace) and carries no URL, so no CRD can point a declared
        agent identity at another endpoint.

        Reading the identity off the URL instead would refuse two
        renderings the controller emits: a global proxy rewrites every
        URL to the proxy host and moves the real one into the
        ``x-kagent-host`` header, and a sandbox agent is reached at the
        controller's own address under `/api/a2a-sandboxes/<ns>/<name>`.
        A hand-written ``config.json`` mounted past the controller can
        still name one agent and reach another — the Known gaps table of
        integrations/kagent/IMPLEMENTATION.md.

        The ``url`` is not read at all here, where the go lane skips a
        url-less entry: this runtime wires every declared remote agent
        whatever its url (``kagent.adk.types.AgentConfig.to_agent``),
        and the go one skips the url-less ones, so each lane's
        inventory follows the tools its own runtime builds.
        """
        name = remote.get("name")
        if not isinstance(name, str):
            raise ConfigRefused(
                f"{path}.name: a remote agent is wired as a tool of its name, and this entry declares none"
            )
        namespace, mark, agent = name.partition(_NAMESPACE_MARK)
        if not mark or not namespace or not agent or _NAMESPACE_MARK in agent:
            raise ConfigRefused(
                f"{path}.name: kagent renders a remote agent as <namespace>__NS__<agent>, and {name!r} "
                "is not that shape"
            )
        namespace, agent = namespace.replace("_", "-"), agent.replace("_", "-")
        if not _SEGMENT.match(namespace) or not _SEGMENT.match(agent):
            raise ConfigRefused(f"{path}.name: the remote agent name {name!r} is outside what the wire can spell")
        self.add(name, agent_spelling(namespace, agent), path)
